src/provision.toit

```diff
@@ -0,0 +1,518 @@
+  static RESP_STATUS_CONNECTED_AUTH_MODE ::= 2 /** type: uint32 */
+  static RESP_STATUS_CONNECTED_SSID ::= 3 /** type: bytes */
+  static RESP_STATUS_CONNECTED_BSSID ::= 4 /** type: bytes */
+  static RESP_STATUS_CONNECTED_CHANNEL ::= 5 /** type: uint32 */
+
+  static SET_CONFIG ::= 12 /** type: message */
+  static SET_CONFIG_SSID ::= 1 /** type: bytes */
+  static SET_CONFIG_PASSWORD ::= 2 /** type: bytes */
+
+  static REQ_APPLY ::= 14 /** type: message */
+
+  ssid/string := ""
+  password/string := ""
+  network := null
+
+  latch/monitor.Latch := ?
+
+  constructor .latch/monitor.Latch:
+
+  set_config r/protobuf.Reader -> ByteArray:
+    r.read_message:
+      r.read_field SET_CONFIG_SSID:
+        ssid = r.read_primitive protobuf.PROTOBUF_TYPE_STRING
+      r.read_field SET_CONFIG_PASSWORD:
+        password = r.read_primitive protobuf.PROTOBUF_TYPE_STRING
+    
+    resp_msg := {
+      MSG: MSG_RESP_CONFIG
+    }
+    resp := protobuf_map_to_bytes_ --message=resp_msg
+    return resp
+
+  set_apply -> ByteArray:
+    network = wifi.open
+        --ssid=ssid
+        --password=password
+
+    resp_msg := {
+        MSG: MSG_RESP_APPY
+    }
+    return protobuf_map_to_bytes_ --message=resp_msg 
+
+  req_status -> ByteArray:
+    ap ::= network.access_point
+
+    resp_msg := {
+      MSG: MSG_RESP_STATUS,
+      RESP_STATUS: {
+        RESP_STATUS_CONNECTED: {
+          RESP_STATUS_CONNECTED_IPV4_ADDR: "$(network.address)",
+          RESP_STATUS_CONNECTED_AUTH_MODE: ap.authmode,
+          RESP_STATUS_CONNECTED_SSID: ap.ssid,
+          RESP_STATUS_CONNECTED_BSSID: ap.bssid,
+          RESP_STATUS_CONNECTED_CHANNEL: ap.channel,
+        }
+      }
+    }
+    return protobuf_map_to_bytes_ --message=resp_msg  
+
+  run data/ByteArray -> ByteArray:
+    resp := #[]
+
+    r := protobuf.Reader data
+    r.read_message:
+      r.read_field MSG:
+        msgid := r.read_primitive protobuf.PROTOBUF_TYPE_INT32
+        if msgid == MSG_SET_APPLY:
+          resp = set_apply
+      r.read_field SET_CONFIG:
+        resp = set_config r
+      r.read_field REQ_STATUS:
+        tmp := r.read_primitive protobuf.PROTOBUF_TYPE_BYTES
+        resp = req_status
+        latch.set true
+      r.read_field REQ_APPLY:
+        tmp := r.read_primitive protobuf.PROTOBUF_TYPE_BYTES
+        resp = set_apply
+
+    return resp
+
+class Provision:
+  service_/BLEService_ := ?
+  security_/Security_ := ?
+  version_task_/Task? := null
+  config_task_/Task? := null
+  session_task_/Task? := null
+  scan_task_/Task? := null
+  latch_ := monitor.Latch
+
+  constructor.ble service_name/string security/Security_:
+    return Provision.ble_with_uuid SERVICE_UUID service_name security
+  
+  constructor.ble_with_uuid service_uuid/ByteArray service_name/string .security_/Security_:
+    service_ = BLEService_ service_uuid service_name
+    add_finalizer this::
+      this.close
+
+  start -> none:
+    if version_task_: throw "Already running"
```

Comment:
Maybe also check that the latch doesn't have a value yet.
Otherwise someone could `start`, `close`, and then `start` again, and we would notice it too late.